experimental/Schemes/SpecOpen.jl

```diff
@@ -0,0 +1,694 @@
+  l = write_as_linear_combination(one(OO(V)), OO(V).(lifted_denominator.(g)))
+  return dot(l, OO(V).(lifted_numerator.(g)))
+end
+
+@Markdown.doc """
+    maximal_extension(
+      X::Spec{BRT, BRET, RT, RET, MST}, 
+      f::AbstractAlgebra.Generic.Frac{RET}
+    ) where {BRT, BRET, RT, RET, MST}
+
+Returns the maximal extension of the restriction of ``f`` 
+to a rational function on ``X`` on a maximal domain of 
+definition ``U ⊂ X``. 
+"""
+function maximal_extension(
+    X::Spec{BRT, BRET, RT, RET, MST}, 
+    f::AbstractAlgebra.Generic.Frac{RET}
+  ) where {BRT, BRET, RT, RET, MST}
+  a = numerator(f)
+  b = denominator(f)
+  W = localized_ring(OO(X))
+  I = quotient(ideal(W, b) + localized_modulus(OO(X)), ideal(W, a))
+  U = SpecOpen(X, I)
+  g = [OO(V)(f) for V in patches(U)]
+  return StructureSheafElem(U, g)
+end
+
+@Markdown.doc """
+    maximal_extension(
+      X::Spec{BRT, BRET, RT, RET, MST}, 
+      f::Vector{AbstractAlgebra.Generic.Frac{RET}}
+    ) where {BRT, BRET, RT, RET, MST}
+
+Returns the extension of the restriction of the ``fᵢ`` as a 
+set of rational functions on ``X`` as *regular* functions to a 
+common maximal domain of definition ``U ⊂ X``.
+"""
+function maximal_extension(
+    X::Spec{BRT, BRET, RT, RET, MST}, 
+    f::Vector{AbstractAlgebra.Generic.Frac{RET}}
+  ) where {BRT, BRET, RT, RET, MST}
+  if length(f) == 0
+    return SpecOpen(X), Vector{StructureSheafElem{BRT, BRET, RT, RET, MST}}()
+  end
+  R = base_ring(parent(f[1]))
+  for a in f
+    R == base_ring(parent(a)) || error("fractions do not belong to the same ring")
+  end
+  R == base_ring(OO(X)) || error("fractions do not belong to the base ring of the scheme")
+  a = numerator.(f)
+  b = denominator.(f)
+  W = localized_ring(OO(X))
+  I = ideal(W, one(W))
+  for p in f
+    I = intersect(quotient(ideal(W, denominator(p)) + localized_modulus(OO(X)), ideal(W, numerator(p))), I)
+  end
+  U = SpecOpen(X, I)
+  return U, [StructureSheafElem(U, [OO(V)(a) for V in patches(U)]) for a in f]
+end
+
+@Markdown.doc """
+    SpecOpenMor{BRT, BRET, RT, RET, MST1, MST2}
+
+Morphisms ``f : U → V`` of open sets ``U ⊂ X`` and ``V ⊂ Y`` of affine schemes.
+These are stored as morphisms ``fᵢ: Uᵢ→ Y`` on the affine patches 
+``Uᵢ`` of ``U``.
+"""
+mutable struct SpecOpenMor{BRT, BRET, RT, RET, MST1, MST2}
+  domain::SpecOpen{BRT, BRET, RT, RET, MST1}
+  codomain::SpecOpen{BRT, BRET, RT, RET, MST2}
+  maps_on_patches::Vector{SpecMor{BRT, BRET, RT, RET, MST1, MST2}}
+
+  # fields used for caching
+  inverse::SpecOpenMor{BRT, BRET, RT, RET, MST2, MST1}
+
+  function SpecOpenMor(
+      U::SpecOpen{BRT, BRET, RT, RET, MST1},
+      V::SpecOpen{BRT, BRET, RT, RET, MST2},
+      f::Vector{SpecMor{BRT, BRET, RT, RET, MST1, MST2}};
+      check::Bool=true
+    ) where {BRT, BRET, RT, RET, MST1, MST2}
+    Y = ambient(V)
+    n = length(f)
+    n == length(patches(U)) || error("number of patches does not coincide with the number of maps")
+    if check
```

Comment:
consider always doing the trivial checks and the expensive checks only at request?

Reply:
Yes. But the other way round, given the default value for check.

Reply:
I think I was not very clear about this: 

The idea is that expensive checks are done by default, once a new object enters the system. They can be switched off by the knowing user in case they want to perform some hard task and know what they're doing. But setting the default to `true` should prevent a lot of hassle. Then, this constructor might be called frequently by internal routines where it is clear that the output objects are correct, once the input was correct. Think, for instance, of multiplying fractions in a localized ring: There is no need to double check that the result is an admissible element of this localization. Here, the situation is similar and all the internal routines can call the constructor with `check=false` to have good performance.

Reply:
Usually I leave the trivial input checks always on, to spot bugs in my code. Only the more expensive (depends on the situation what is considered as expensive) checks to be turned off at request (default all checks are on).

Reply:
Ok, same here. But everything within the `if check`-block is indeed rather expensive in this case.